apps/prairielearn/src/tests/groupExam.test.js

```diff
@@ -0,0 +1,401 @@
+    __csrf_token: csrfToken,
+    groupName: groupName,
+  };
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams(form),
+  });
+  assert.isOk(res.ok);
+  locals.$ = cheerio.load(await res.text());
+};
+
+/**
+ * Joins group as current user with CSRF token and loads page with cheerio.
+ * @param {String} assessmentUrl
+ * @param {String} joinCode
+ */
+const joinGroup = async (assessmentUrl, joinCode) => {
+  const form = {
+    __action: 'join_group',
+    __csrf_token: locals.__csrf_token,
+    join_code: joinCode,
+  };
+  const res = await fetch(assessmentUrl, {
+    method: 'POST',
+    body: new URLSearchParams(form),
+  });
+  assert.isOk(res.ok);
+  locals.$ = cheerio.load(await res.text());
+};
+
+describe('Group based exam assess control on student side', function () {
+  this.timeout(20000);
+  before('set authenticated user', function (callback) {
+    storedConfig.authUid = config.authUid;
+    storedConfig.authName = config.authName;
+    storedConfig.authUin = config.authUin;
+    callback(null);
+  });
+  before('set up testing server', helperServer.before(TEST_COURSE_PATH));
+  after('shut down testing server', helperServer.after);
+  after('unset authenticated user', function (callback) {
+    Object.assign(config, storedConfig);
+    callback(null);
+  });
+
+  step('1. the database contains a group-based exam assessment', async function () {
+    const result = await sqldb.queryAsync(sql.select_group_work_exam_assessment, []);
+    assert.lengthOf(result.rows, 2);
+    assert.notEqual(result.rows[0].id, undefined);
+    locals.assessment_id = result.rows[0].id;
+    locals.assessmentUrl = locals.courseInstanceUrl + '/assessment/' + locals.assessment_id;
+    locals.instructorAssessmentsUrlGroupTab =
+      locals.courseInstanceUrl + '/instructor/assessment/' + locals.assessment_id + '/groups';
+    locals.assessment_id_2 = idsEqual(result.rows[1].id, locals.assessment_id)
+      ? result.rows[0].id
+      : result.rows[1].id;
+    locals.assessmentUrl_2 = locals.courseInstanceUrl + '/assessment/' + locals.assessment_id_2;
+  });
+
+  step(
+    '2. GET to instructor assessments URL group tab for the first assessment loads correctly',
+    async function () {
+      // should load successfully
+      const res = await fetch(locals.instructorAssessmentsUrlGroupTab);
+      page = await res.text();
+
+      // should parse
+      locals.$ = cheerio.load(page);
+
+      // check for CSRF tokens
+      elemList = locals.$('form input[name="__csrf_token"]');
+      assert.lengthOf(elemList, 5);
+      // there are 6 occurrences of the same csrf, we will pick the first one
+      assert.nestedProperty(elemList[0], 'attribs.value');
+      locals.__csrf_token = elemList[0].attribs.value;
+      assert.isString(locals.__csrf_token);
+    },
+  );
+
+  step('3. Group config in database is correct', async function () {
+    const params = {
+      assessment_id: locals.assessment_id,
+    };
```

Comment:
Nit: we've been leaning towards inlining params in the function call instead of declaring a separate variable, especially when the params aren't referenced or reused in multiple places.